### PR TITLE
Plat 606 openhim mediators refactor

### PR DESCRIPTION
**Pull request by nour-borgi**

### CODE REVIEW

File: interoperability-layer-openhim/docker-compose-mongo.cluster.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ services:
   mongo-2:
     image: mongo:4.2
     volumes:
-      - 'openhim-mongo2:/data/db'
+      - 'openhim-mongo-2:/data/db'
```
Comment (marrouchi) [inline interoperability-layer-openhim/docker-compose-mongo.cluster.yml:14]:
```suggestion
      - 'openhim-mongo-02:/data/db'
```

File: interoperability-layer-openhim/swarm.sh [hunk 2]
```diff
@@ -1,52 +1,63 @@
...
+
+  mongo_services=(
+    "mongo-1"
+  )
+  if [[ "${NODE_MODE}" == "cluster" ]]; then
+    mongo_services=(
+      "${mongo_services[@]}"
+      "mongo-2"
+      "mongo-3"
+    )
```
Comment (marrouchi) [inline interoperability-layer-openhim/swarm.sh:29]: How about for i in {1..3}

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: interoperability-layer-openhim/swarm.sh (hunk 2 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: done in [c0d70ba](https://github.com/jembi/platform/pull/197/commits/c0d70ba7882b95805a6b144f3bbf99e7b60f6c01)

File: interoperability-layer-openhim/docker-compose-mongo.cluster.yml (hunk 1 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/docker-compose-mongo.cluster.yml]: done in [c0d70ba](https://github.com/jembi/platform/pull/197/commits/c0d70ba7882b95805a6b144f3bbf99e7b60f6c01)

**Review by marrouchi — APPROVED** [review]

File: interoperability-layer-openhim/swarm.sh [hunk 3]
```diff
@@ -1,52 +1,64 @@
...
+  MODE=$2
+
+  COMPOSE_FILE_PATH=$(
+    cd "$(dirname "${BASH_SOURCE[0]}")" || exit
+    pwd -P
+  )
+
+  UTILS_PATH="${COMPOSE_FILE_PATH}/../utils"
+
+  mongo_services=(
```
Comment (marrouchi) [inline interoperability-layer-openhim/swarm.sh:21]: Since it's initialized here, we will get duplicate after the loop in the cluster mode.

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: interoperability-layer-openhim/swarm.sh (hunk 3 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh:21]: Nice catch :) done

File: interoperability-layer-openhim/swarm.sh [hunk 4]
```diff
@@ -1,52 +1,67 @@
...
+# shellcheck disable=SC1091
+function import_sources() {
+  source "${UTILS_PATH}/docker-utils.sh"
+  source "${UTILS_PATH}/config-utils.sh"
+  source "${UTILS_PATH}/log.sh"
 }
 
-verify_mongos() {
-  log info 'Waiting to ensure all the mongo instances for the replica set are up and running'
+function verify_mongos() {
```
Comment (Mark-Labuschagne) [inline interoperability-layer-openhim/swarm.sh:57]: Is verify_mongos not rendered redundant by docker::deploy_sanity?

File: interoperability-layer-openhim/swarm.sh [hunk 5]
```diff
@@ -60,112 +75,119 @@ verify_mongos() {
...
+    docker::deploy_service "${COMPOSE_FILE_PATH}" "docker-compose-mongo.yml" "$MONGO_CLUSTER_COMPOSE_FILENAME" "$MONGO_DEV_COMPOSE_FILENAME"
+    docker::deploy_sanity "${mongo_services[@]}"
 
-    try "docker stack deploy -c ${COMPOSE_FILE_PATH}/importer/docker-compose.config.yml instant" "Failed to deploy config importer"
+    verify_mongos
+    prepare_console_config
 
-    log info "Waiting to give core config importer time to run before cleaning up service"
+    docker::deploy_service "${COMPOSE_FILE_PATH}" "docker-compose.yml" "docker-compose.stack-1.yml" "$OPENHIM_DEV_COMPOSE_FILENAME"
+    docker::deploy_sanity "openhim-core" "openhim-console"
```
Comment (Mark-Labuschagne) [inline interoperability-layer-openhim/swarm.sh:120]: Why don't we just scale the services up doing docker service scale?

File: interoperability-layer-openhim/swarm.sh [hunk 6]
```diff
@@ -60,112 +75,119 @@ verify_mongos() {
...
+  ) || {
+    log error "Failed to scale up Interoperability Layer OpenHIM package"
+    exit 1
+  }
+}
 
-    config::remove_config_importer interoperability-layer-openhim-config-importer
+function start_package() {
```
Comment (Mark-Labuschagne) [inline interoperability-layer-openhim/swarm.sh:127]: This entire function can be removed. Put the variable things into the init_vars function, then move the rest into the main function by splitting out the `init` and `up` sections

File: interoperability-layer-openhim/swarm.sh [hunk 7]
```diff
@@ -60,112 +75,119 @@ verify_mongos() {
...
-    local mongo_dev_compose_param=""
-    local openhim_dev_compose_param=""
-  fi
-
-  if [[ "${ACTION}" == "init" ]]; then
-    config::set_config_digests "${COMPOSE_FILE_PATH}"/docker-compose.yml
-    config::set_config_digests "${COMPOSE_FILE_PATH}"/importer/docker-compose.config.yml
-
-    try "docker stack deploy -c ${COMPOSE_FILE_PATH}/docker-compose-mongo.yml $mongo_cluster_compose_param $mongo_dev_compose_param instant" "Failed to deploy mongo"
+function init_package() {
```
Comment (Mark-Labuschagne) [inline interoperability-layer-openhim/swarm.sh:83]: Please be consistent with other packages, choose one, either `init_package` or `initialize_package`

File: interoperability-layer-openhim/swarm.sh (hunk 4 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: yes nice done :)

File: interoperability-layer-openhim/swarm.sh (hunk 7 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: cool done

File: interoperability-layer-openhim/swarm.sh [hunk 8]
```diff
@@ -1,171 +1,151 @@
...
-    config::set_config_digests "${COMPOSE_FILE_PATH}"/docker-compose.yml
-    config::set_config_digests "${COMPOSE_FILE_PATH}"/importer/docker-compose.config.yml
+  if [[ "${NODE_MODE}" == "cluster" ]]; then
+    mongo_cluster_compose_filename="docker-compose-mongo.cluster.yml"
+  fi
 
-    try "docker stack deploy -c ${COMPOSE_FILE_PATH}/docker-compose-mongo.yml $mongo_cluster_compose_param $mongo_dev_compose_param instant" "Failed to deploy mongo"
+  (
+    docker::deploy_service "${COMPOSE_FILE_PATH}" "docker-compose-mongo.yml" "$mongo_cluster_compose_filename" "$mongo_dev_compose_filename"
+    docker::deploy_sanity "${mongo_services[@]}"
```
Comment (marrouchi) [inline interoperability-layer-openhim/swarm.sh:83]: Should the deploy sanity be performed after initiating the replica set ?

File: interoperability-layer-openhim/swarm.sh [hunk 9]
```diff
@@ -1,171 +1,151 @@
...
 
-    if [[ "${STATEFUL_NODES}" == "cluster" ]]; then
-      try "${COMPOSE_FILE_PATH}/initiateReplicaSet.sh" "Fatal: Initate Mongo replica set failed."
+    if [[ "${NODE_MODE}" == "cluster" ]]; then
+      try "${COMPOSE_FILE_PATH}/initiate-replica-set.sh" throw "Fatal: Initiate Mongo replica set failed"
```
Comment (marrouchi) [inline interoperability-layer-openhim/swarm.sh:86]: Initiating the replicaset should be performed when the action = "init"

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: interoperability-layer-openhim/swarm.sh (hunk 5 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: I removed all of these, init will do both 👍🏼

File: interoperability-layer-openhim/swarm.sh (hunk 6 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: cool done :)

File: interoperability-layer-openhim/swarm.sh (hunk 8 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh]: done :)

File: interoperability-layer-openhim/swarm.sh (hunk 9 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/swarm.sh:85]: done :)

File: interoperability-layer-openhim/package-metadata.json [hunk 10]
```diff
@@ -6,7 +6,7 @@
   "type": "infrastructure",
   "dependencies": [],
   "environmentVariables": {
-    "STATEFUL_NODES": "single",
+    "NODE_MODE": "single",
     "MONGO_SET_COUNT": "1",
```
Comment (marrouchi) [inline interoperability-layer-openhim/package-metadata.json:10]: We don't need MONGO_SET_COUNT anymore, right ?

**Review by marrouchi — APPROVED** [review]
Looks good with 1 minor fix.

File: interoperability-layer-openhim/package-metadata.json (hunk 10 above)
Comment (nour-borgi) [reply interoperability-layer-openhim/package-metadata.json:10]: we still need it in the script initiate-replica-set.sh |this script needs to be refactored as well]

**Review by marrouchi — APPROVED** [review]

**Review by Mark-Labuschagne — APPROVED** [review]
NICE!!

**Pull request merged**